### DISCUSSION TITLE
Try to understand how Babel work in TSConfig documentation for option `allowSyntheticDefaultImports`

### DISCUSSION THREAD

**Discussion by nyngwang** [babel/babel#16335] [Q&A] [answered]
From their explanation: [This is the line](https://www.typescriptlang.org/tsconfig#:~:text=this%20code%20raises%20an%20error%20because%20there%20isn%E2%80%99t%20a%20default%20object%20which%20you%20can%20import.%20even%20though%20it%20feels%20like%20it%20should.%20for%20convenience%2C%20transpilers%20like%20babel%20will%20automatically%20create%20a%20default%20if%20one%20isn%E2%80%99t%20created.)

I don't know which plugin does what it means by "_Babel will automatically create a default if one isn’t created_". I know that we have preset `@babel/preset-typescript` but is it really the plugin that is responsible for this transformation? I'm not sure, but I want to know.

> This code raises an error because there isn’t a default object which you can import. Even though it feels like it should. For convenience, transpilers like **Babel will automatically create a default** if one isn’t created.

```js
// @filename: utilFunctions.js
const getStringLength = (str) => str.length;
 
module.exports = {
  getStringLength,
};
 
// @filename: index.ts
import utils from "./utilFunctions";
 
const count = utils.getStringLength("Check JS");
```
The error:
```err
Module '"/home/runner/work/TypeScript-Website/TypeScript-Website/utilFunctions"' has no default export.
```

**Comment by nicolo-ribaudo** [comment, marked as answer]
It's `@babel/plugin-transform-modules-commonjs`. It's not just Babel btw, every tool does that when compiling ESM to CJS :) Also Node.js does the same.

**Reply by nyngwang** [reply]
@nicolo-ribaudo Did you mean that I don't need to worry about installing some custom plugins (OK now I know its name `@babel/plugin-transform-modules-commonjs`) to make it work? I was worried about whether I have to install some plugins in addition to `@babel/preset-typescript` to make sure that every option I might use in TSConfig can work correctly to simulate what Babel actually does. (my current understanding is that they're trying to simulate how bundler&transpiler work.)

Great I found that `@babel/preset-typescript` actually [includes it](https://github.com/babel/babel/blob/5ca9b9a84f5f06d16704c53717c9baae2b347839/packages/babel-preset-typescript/package.json#L23) :)

**Reply by liuxingbaoyu** [reply]
I think you still need to install it, the dependency will only be applied to the `.cts` file.

**Reply by nyngwang** [reply]
@nicolo-ribaudo  On the other hand, I'm trying to understand the purpose of the option `strict`. It tries to convey whether we should only accept native(so it's strict) ESM modules instead of a non-standard(or maybe there is a better adj.) one with `__esModule: true`, correct?

@liuxingbaoyu Could you elaborate more on it?

**Reply by liuxingbaoyu** [reply]
https://github.com/babel/babel/blob/5ca9b9a84f5f06d16704c53717c9baae2b347839/packages/babel-preset-typescript/src/index.ts#L83-L92

It is only used for `.cts` files. If you want to output `.ts` files to cjs, you still need to add `@babel/plugin-transform-modules-commonjs` manually.

**Reply by nyngwang** [reply]
@liuxingbaoyu Hi, thank you for the point. I didn't think about the target I wanted with `@babel/preset-typescript`. I also found that `@babel/preset-env` [covers it](https://github.com/babel/babel/blob/5ca9b9a84f5f06d16704c53717c9baae2b347839/packages/babel-preset-env/package.json#L67).

On the other hand, what do you think about my confusion in the first comment?

> I was worried about whether I have to install some plugins in addition to @babel/preset-typescript to make sure that every option I might use in TSConfig can work correctly to simulate what Babel actually does.

**Reply by liuxingbaoyu** [reply]
Yes, this is necessary. A ts option may mean one or more babel plugins, but they are generally contained in `@babel/preset-env`.
